강원대 BE 손상욱 - step1 유효성 검사, 예외 처리

thstkddnr20 opened the pull request:
## 주요 변경사항
생성, 수정 dto 유효성 검증
ProductApiController e2e 테스트 코드 작성

## 맞닥뜨린 문제
1. 테스트마다 schema.sql의 테이블 생성 코드가 적용되어 다음 클래스의 테스트가 실행이 되지 않는 문제
 이전에 ProductServiceTest를 작성한 상태였고 이번에 새로운 ProductApiControllerTest를 작성하였는데, 테스트가 잘 작동하는지 확인하기 위해 각각의 테스트를 별도로 실행시키면 문제가 없지만 test 디렉토리 안에있는 테스트를 한꺼번에 실행시킬 경우에 schema.sql의 테이블 생성 코드가 다음 테스트 클래스를 실행시킬 때도 중복 적용되어 테스트에 실패하는 문제가 있었습니다. 이를 해결하기 위해 dropTable.sql을 작성하여 테스트 클래스의 작동이 모두 종료된 후 다음 클래스로 넘어가기 전에 drop table을 해주는 방식으로 해결하였습니다.

2. delete문은 id가 없어도 올바르게 작동하는 문제
ProductApiControllerTest의 test7() 메서드("없는 아이디의 상품 삭제를 요청할 경우 상태코드 400을 반환한다.")를 작성하던 중, ProductRepository의 deleteById() 메서드에서 존재하지 않은 id가 들어왔을 경우에도 오류를 반환하지 않고 정상적으로 작동하는 문제가 있었습니다. 이를 해결하기 위해 아래와 같은 방식으로 변경하였습니다.
    ```java
    public void deleteByid(Long id) {
        String find = "select * from product where id=?";
        try {
            jdbcTemplate.queryForObject(find, toProduct(), id);
        } catch (EmptyResultDataAccessException e) {
            throw new ProductNotFoundException(id);
        }
        String sql = "delete from product where id=?";
        jdbcTemplate.update(sql, id);
    }    
    ```
    조회쿼리는 id가 없을 경우 예외를 던지므로 해당 id에 대해서 조회쿼리를 먼저 날린 후, catch 문에서 예외를 잡아 직접                                  ProductNotFoundException을 던지도록 하였습니다. 예외는 다시 Advice에서 잡혀 400 bad request를 반환하도록 하였습니다.

    원하는 흐름대로 서비스가 작동하기 위해서 테스트 코드를 작성하는 것이 중요하다는 것을 다시 느끼게 된 부분인 것 같습니다.

## 중점적으로 리뷰받고 싶은 곳
1. 맞닥뜨린 문제의 1번에서 더 나은 방법이 없을 지 의논하고 싶습니다!
2. 상품이름에 카카오가 포함되었을 경우도 dto에 대한 올바르지 않은 요청이라고 판단하여 똑같이 400 bad request를 반환하였습니다. 이 부분에 대하여 어떻게 생각하시는지 궁금합니다!

Commit:
docs: README.md step1 구현 기능 작성

Commit:
feat: CreateProductRequest, UpdateProductRequest 유효성 검증 추가

Commit:
test: ProductApiController e2e 테스트 작성
상품 삭제에서 id가 없는 경우에도 repository에서 에러가 발생하지 않는 문제가 있어 의도적으로 exception을 던지도록 하였습니다.

Commit:
fix: dropTable.sql 테스트 코드 적용
전체 테스트를 한번에 실행할 경우 다음 테스트 클래스로 넘어갈 때 다시 schema.sql의 create table product를 실행하여 테스트가 실패, 테스트 클래스 단위가 넘어갈 때마다 drop table하여 해결

leaf-upper commented on src/main/resources/dropTable.sql:
```diff
@@ -0,0 +1 @@
+drop table product;
```
1번문제에 대해서 간략하게 제 생각을 말씀드리자면,
테스트마다 독립적인 실행환경을 구성하는 방식은 좋습니다.
현재는 h2 db를 쓰고계시니, 테스트별로 테이블을 생성하고, 테이블을 삭제하는것도
합리적이라고 생각이 들어요. 
매 테스트마다 이미 기존재하는 테이블을 생성하여 오류가 발생한다면
"create table if not exists" 을 사용하실수도 있고, 테스트가 종료된후 테이블을 제거하는게 아니라, 생성한 데이터를 제거하는 방식으로도 테스트를 하실수 있을것 같아요.
여러가지 방식이 있겠지만 현재 방식도 저는 좋은것 같아요!

leaf-upper commented on src/main/java/gift/dto/product/CreateProductRequest.java:
```diff
@@ -1,12 +1,23 @@
-import jakarta.validation.constraints.NotNull;
+import jakarta.validation.constraints.*;
 
 public record CreateProductRequest(
```
확실히 Controller 계층에서는, Validation을 확실히 해주는게 좋습니다.
Controller 계층에서 Vadliation을 제대로 수행했다면 이후 도메인계층 (Service, Repository)
에서는 이 데이터를 믿어갈수 있을것 같아요.

leaf-upper commented on src/main/java/gift/dto/product/CreateProductRequest.java:
```diff
@@ -1,12 +1,23 @@
+        @Max(value = 100_000_000, message = "수량은 1억개 이하여야 합니다.")
+        Integer quantity
 ) {
 
```
불변객체인만큼 empty()에 대해서 매번 객체를 생성하는게 아니라, 한개의 객체를 만들어서 재사용해도 될것 같아요

leaf-upper commented on src/main/java/gift/dto/product/UpdateProductRequest.java:
```diff
@@ -1,12 +1,23 @@
+        @Max(value = 100_000_000, message = "수량은 1억개 이하여야 합니다.")
+        Integer quantity
 ) {
     public static UpdateProductRequest from(ProductManageResponse response) {
```
CreateProductRequest에서 드린 코멘트와 동일합니다.

thstkddnr20 replied:
안녕하세요 멘토님 코멘트 남겨주신 것 중 from() 부분에서 궁금한 점이 생겨 답글 남겼습니다!

위 코멘트인 empty()가 문제 없이 작동 할 것이라는 것은 이해가 됐습니다.
`private static final CreateProductRequest EMPTY = new CreateProductRequest(null, null, null);`

캐싱된 객체를 empty() 메서드로 반환 -> model로 전달 -> 사용자의 입력 받아오기 -> 받아온 입력은 기존 EMPTY 객체와 다른 인스턴스의 새로운 dto의 필드에 할당될 것
이 순서대로 진행되므로 문제가 없음을 깨달았습니다.

하지만 from() 메서드 같은 경우는 파라미터로 ProductManageResponse를 받아오는 상태이며, 모든 필드가 final로 선언되어있는 record 클래스에서 기존의 캐싱된 객체에 인자로 들어온 필드를 바인딩하는 것이 불가능 하다고 생각되었습니다.

필드값을 할당하기 위해서는 생성자를 사용해야 함 -> 캐싱된 객체의 필드값을 변경하는 것은 불가능

그렇다고 이를 class로 다시 변경하기에는 mutable 상태가 되어 여러 쓰레드에서 공유하는 것이 위험하다고 생각이 들었습니다.
이 부분에 대해서 조금 더 자세히 알려주실 수 있을까요?

leaf-upper replied:
@thstkddnr20 
안녕하세요 상욱님
제가 코멘트를 잘못남겨드렸습니다. 혹시 이 코멘트로 너무많은 고민을 하셨을까봐 죄송하네요.. 
상욱님이 말씀해주신게 맞아요 from() 메서드는 캐싱하기 어렵습니다 ㅎㅎ..

thstkddnr20 replied:
안녕하세요 멘토님! dto를 늘 new 키워드를 사용하여 새로운 객체를 반환하는 방식으로만 사용했었는데, 이번 기회로 객체의 생명주기를 이해하고 새로운 방식으로 사용할 수 있음을 깨닫게 되었습니다. 항상 감사합니다!!!

leaf-upper commented on src/main/java/gift/repository/ProductRepository.java:
```diff
@@ -58,6 +59,12 @@ public Optional<Product> findById(Long id) {
     }
 
     public void deleteByid(Long id) {
```
camelCase를 지켜주세요 ㅎㅎ

leaf-upper commented on src/main/java/gift/repository/ProductRepository.java:
```diff
@@ -58,6 +59,12 @@ public Optional<Product> findById(Long id) {
     }
 
     public void deleteByid(Long id) {
+        String find = "select * from product where id=?";
```
delete은 영향받는 Row의 개수를 Return할수있어요
id를 기반으로 삭제한다고 했을때 id는 PK이기 때문에
데이터베이스에 0개 혹은 1개만 있어야합니다.
1개인경우 정상 삭제 되었다고 볼수있고, 0개라면 정상삭제되지 않았다고 볼 수 있지않을까요?

leaf-upper commented on src/test/java/gift/service/ProductServiceTest.java:
```diff
@@ -17,7 +17,8 @@
 @SpringBootTest
-@Sql(scripts = "/cleanup.sql", executionPhase = Sql.ExecutionPhase.BEFORE_TEST_METHOD)
+@Sql(scripts = "/cleanup.sql", executionPhase = Sql.ExecutionPhase.AFTER_TEST_METHOD)
+@Sql(scripts = "/dropTable.sql", executionPhase = Sql.ExecutionPhase.AFTER_TEST_CLASS)
```
테스트가 끝난뒤 테이블을 제거하는작업도 필요할수있다고 생각해요~

thstkddnr20 replied:
dropTable.sql에 `drop table product;`가 AFTER_TEST_CLASS로 호출되어 테이블 제거하고 있습니다!

leaf-upper reviewed (changes requested):
안녕하세요 상욱님~
Step1 코드 잘 작성해주신것 같습니다.
몇가지 코멘트를 남겨드렸는데요~ 코멘트 보시고
수정사항 반영해주시고 다시 리뷰요청 부탁드려요!
고생하셨습니다!

Commit:
refactor: 리뷰 반영하여 코드 수정
1. CreateProductRequest 클래스 static final EMPTY 도입
2. ProductRepository의 deleteById() camelCase 준수, 삭제 로직 영향받은 컬럼을 확인하여 에러 처리
3. 불필요한 주석 제거

thstkddnr20 commented:
## 주요 변경 사항
리뷰 반영하여 수정을 진행하였습니다
1. CreateProductRequest 클래스에 static final EMPTY를 도입하여 매번 새로운 객체를 만들지 않고 dto 반환
2. ProductRepository의 deleteById() 메서드 camelCase 준수, 삭제 로직 영향 받은 컬럼을 확인하여 에러 처리
3. 불필요한 주석 제거

## 맞닥뜨린 문제
UpdateProductRequest를 CreateProductRequest 방식과 비슷하게 풀어보려고 해보았으나 개인적으로는 해결 방법을 찾지 못하였습니다.
위 답글에 문제에 대한 내용이 작성되어 있습니다!

leaf-upper commented on src/main/java/gift/service/ProductManageService.java:
```diff
@@ -39,7 +39,7 @@ private Product getById(Long id) {
     }
 
     public void deleteProduct(Long id) {
-        productRepository.deleteByid(id);
```
👍 👍  camelCase로 잘바꿔주셨네요~

leaf-upper reviewed (approved):
👍 👍 코멘트 남겨드린 부분 잘 수정해주신거 확인했어요~
제가 리뷰를 잘못 남겨드린 부분이 있는데 그 부분에 대해서 많이 고민해주신것 같네요
상욱님이 코멘트로 주신말이 맞습니다! 너무 잘하셨어요
다음 Step진행해주셔도 좋을것 같아서 어프로브드립니다 고생하셨어요!
